File: python/loader.py
import numpy as np

_SONAR = 'dataset/sonar/sonar.csv'

class Loader:

    def __init__(self, dataset='sonar', tau=1e-2):
        self.loaded = False
        if(dataset == 'sonar'):
            data = np.genfromtxt(_SONAR, delimiter=',')
            self.X = data[:,:-1] # delete the last column
            label = np.genfromtxt(_SONAR, dtype='S1', usecols=self.X.shape[1], delimiter=',')
            self.y = (label != label[0]).astype(int) # make labels either zeros or ones
            self.loaded = True
            self.tau = tau
        return

    def get_function(self):
        assert self.loaded

        X = self.X
        y = self.y

        # get f lambda function
        f = lambda w : np.sum(np.log(1+np.exp(-np.matmul(np.diag(y), np.matmul(X, w)))))+0.5*self.tau*np.dot(w,w)

        # get grad_f lambda function
        h = lambda w: np.exp(-np.matmul(np.diag(y), np.matmul(X, w)))
        grad_num = lambda w : np.matmul(np.diag(h(w)), np.matmul(np.diag(y), X))
        grad_den = lambda w : np.linalg.inv(np.eye(y.shape[0]) + np.diag(h(w)))

        grad_f = lambda w : -np.sum(np.matmul(grad_den(w), grad_num(w)), 0)+self.tau*w

        input_size = X.shape[1]

        return f, grad_f, input_size

    def get_constants(self):
        assert self.loaded

        X = self.X
        y = self.y
        sigma = self.tau
        L = 0.25*(np.linalg.norm(X, ord='fro')**2)+self.tau

        return sigma, L

File: matlab/loader.m
function [f,grad_f,f_best,n,m,sigma,L] = loader(tau)

filename = 'dataset/sonar/sonar.csv';
data = readtable(filename);

X = table2array(data(:,1:end-1));
y = table2array(data(:,end));
y = [y{:}] ~= 'R';
%y = 1- y;

n = size(X,1);
m = size(X,2);

f = @(w) sum(log(1+exp(-diag(y)*X*w)))+0.5*tau*(w'*w);

h = @(w) exp(-diag(y)*X*w);
grad_f = @(w) -sum((eye(n)+diag(h(w))) \ (diag(h(w))*diag(y)*X) ) + tau*w';

sigma = tau;
L = 0.25*norm(X,'fro')^2+tau;

% f_best is found running the algorithms for 10000 iterations
if(tau == 1e1)
    f_best = 78.172065242606990;
elseif(tau == 1e-1)
    f_best = 67.667820745302360;
elseif(tau == 1e-3)
    f_best = 67.245844716666310;
else % tau == 1e-6
    f_best = 67.235303085186970;
end

end

